lib/participant.js

```diff
@@ -160,6 +211,30 @@ class Participant extends EventEmitter {
     return track;
   }
 
+
+  /**
+   * @private
+   */
+  _addTrackPublication(publication) {
+    const log = this._log;
+    if (this.trackPublications.has(publication.trackSid)) {
+      return null;
+    }
+    this.trackPublications.set(publication.trackSid, publication);
+
+    const trackPublicationsByKind = {
+      audio: this.audioTrackPublications,
+      data: this.dataTrackPublications,
+      video: this.videoTrackPublications
+    }[publication.kind];
+    trackPublicationsByKind.set(publication.trackSid, publication);
+    reemitTrackPublicationEvents(this, publication);
```

Comment:
Here's an idea for 2.x:

1. We maintain a single, mutable `_trackPublications` Map internally.
2. We introduce a read-only Map class that  wraps mutable Maps.
3. We update `trackPublications` to be a read-only Map wrapping `_trackPublications`
4. We introduce some mechanism such that `videoTrackPublications` and `audioTrackPublications` are just read-only "views" into `trackPublications`/`_trackPublications`.

3 would solve the problem of users theoretically being able to mutate our Maps; 4 would let us simplify the bookkeeping here. Just a thought...